### PR TITLE
Implement unsizing in the new trait solver

### PR DESCRIPTION
**Pull request by compiler-errors**
This makes hello world compile! Ignore the first commit, that's just #107146 which is waiting on merge.

I'll leave some comments inline about design choices that might be debatable.

r? @lcnr (until we have a new trait solver reviewer group...)

### CODE REVIEW

File: compiler/rustc_trait_selection/src/solve/trait_goals.rs [hunk 1]
```diff
@@ -185,6 +186,206 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
...
+                }
+                _ => Err(NoSolution),
+            }
+        })
+    }
+
+    fn consider_builtin_dyn_unsize_candidates(
+        ecx: &mut EvalCtxt<'_, 'tcx>,
+        goal: Goal<'tcx, Self>,
+    ) -> Vec<CanonicalResponse<'tcx>> {
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/trait_goals.rs:318]: This currently returns one response for every supertrait that could be unsized. 

E.g. given:

```
trait Foo: Bar<i32> + Bar<u32> {}
```

Then for `dyn Foo: Unsize<Bar<_>>`, we would have to provide two responses. 

Alternatively, we could just return *one* response (ambiguity) and merge this back into `consider_builtin_unsize_candidate`? 

Not sure if that has implications on inference or when we have two identical candidates, modulo normalization, perhaps, like `trait Foo: Bar<i32> + Bar<<<i32 as Identity>::Identity>`.

File: compiler/rustc_trait_selection/src/solve/trait_goals.rs [hunk 2]
```diff
@@ -185,6 +186,206 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
...
+                (&ty::Dynamic(_, _, ty::Dyn), &ty::Dynamic(_, _, ty::Dyn)) => {
+                    // Dyn upcasting is handled separately, since due to upcasting,
+                    // when there are two supertraits that differ by substs, we
+                    // may return more than one query response.
+                    return Err(NoSolution);
+                }
+                // `T` -> `dyn Trait` unsizing
+                (_, &ty::Dynamic(data, region, ty::Dyn)) => {
+                    // Can only unsize to an object-safe type
+                    // FIXME: Can auto traits be *not* object safe?
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/trait_goals.rs:212]: i don't think so :grin:

File: compiler/rustc_trait_selection/src/solve/project_goals.rs [hunk 3]
```diff
@@ -391,6 +391,20 @@ impl<'tcx> assembly::GoalKind<'tcx> for ProjectionPredicate<'tcx> {
...
     }
+
+    fn consider_builtin_unsize_candidate(
+        _ecx: &mut EvalCtxt<'_, 'tcx>,
+        goal: Goal<'tcx, Self>,
+    ) -> QueryResult<'tcx> {
+        bug!("`Unsize` does not have an associated type: {:?}", goal);
+    }
+
+    fn consider_builtin_dyn_unsize_candidates(
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/project_goals.rs:402]: `dyn_upcast_candidates`. We don't really unsize anything here, do we?

File: compiler/rustc_trait_selection/src/solve/trait_goals.rs [hunk 4]
```diff
@@ -185,6 +186,206 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
...
+                let new_a_data = tcx.mk_poly_existential_predicates(new_a_data);
+                let new_a_ty = tcx.mk_dynamic(new_a_data, b_region, ty::Dyn);
+
+                // We also require that A's lifetime outlives B's lifetime.
+                let mut nested_obligations = ecx.infcx.eq(goal.param_env, new_a_ty, b_ty)?;
+                nested_obligations.push(
+                    goal.with(tcx, ty::Binder::dummy(ty::OutlivesPredicate(a_region, b_region))),
+                );
+
+                responses.push(ecx.evaluate_all_and_make_canonical_response(nested_obligations)?);
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/trait_goals.rs:364]: could you return `ecx.evaluate_all_and_make_canonical_response(nested_obligations)` and let the caller do `responses.push(unsize_dyn_to_principal(Some(erased_trait_ref));`?

File: compiler/rustc_ty_utils/src/ty.rs [hunk 5]
```diff
@@ -407,6 +408,56 @@ fn asyncness(tcx: TyCtxt<'_>, def_id: DefId) -> hir::IsAsync {
...
+
+        ty::GenericArgKind::Const(ct) => match ct.kind() {
+            ty::ConstKind::Param(p) => Some(p.index),
+            _ => None,
+        },
+    };
+
+    // FIXME(eddyb) cache this (including computing `unsizing_params`)
+    // by putting it in a query; it would only need the `DefId` as it
+    // looks at declared field types, not anything substituted.
```
Comment (lcnr) [inline compiler/rustc_ty_utils/src/ty.rs:432]: ? :heart:

**Review by lcnr — COMMENTED** [review]
r=me after nits ^^

this impl looks correct but I didn't check that it perfectly matches the existing one. Will just get test failures if there are differences cause I don't think this is unsound.

File: compiler/rustc_ty_utils/src/ty.rs (hunk 5 above)
Comment (compiler-errors) [reply compiler/rustc_ty_utils/src/ty.rs:432]: why did this comment return 🤔

Comment (compiler-errors) [reply compiler/rustc_ty_utils/src/ty.rs:432]: oh, lol, never removed it from the  code when i query-ified things lol

**Comment by compiler-errors** [comment]
Waiting on #107146 to land.

**Pull request merged**